contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,419 @@
+                          latched_conn.streamInfo(), {});       
+}
+
+TcpUpstream::TcpUpstream(Router::UpstreamToDownstream* upstream_request,
+                         Envoy::Tcp::ConnectionPool::ConnectionDataPtr&& upstream, Dso::TcpUpstreamDsoPtr dynamic_lib,
+                         FilterConfigSharedPtr config)
+    :route_entry_(upstream_request->route().routeEntry()), upstream_request_(upstream_request), upstream_conn_data_(std::move(upstream)), dynamic_lib_(dynamic_lib),
+    config_(config), req_(new RequestInternal(*this)), encoding_state_(req_->encodingState()), decoding_state_(req_->decodingState()) {
+  
+  // req is used by go, so need to use raw memory and then it is safe to release at the gc
+  // finalize phase of the go object.
+  req_->plugin_name.data = config_->pluginName().data();
+  req_->plugin_name.len = config_->pluginName().length();
+
+  upstream_conn_data_->connection().enableHalfClose(true);
+  upstream_conn_data_->addUpstreamCallbacks(*this);
+}
+
+TcpUpstream::~TcpUpstream() {
+  auto reason = (decoding_state_.isProcessingInGo() || encoding_state_.isProcessingInGo())
+                  ? DestroyReason::Terminate
+                  : DestroyReason::Normal;
+
+  dynamic_lib_->envoyGoOnTcpUpstreamDestroy(req_, int(reason));
+}
+
+bool TcpUpstream::initRequest() {
+  if (req_->configId == 0) {
+    req_->setWeakFilter(weak_from_this());
+    req_->configId = config_->getConfigId();
+    return true;
+  }
+  return false;
+}
+
+void TcpUpstream::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "tcp upstream encodeData: {}", data.toString());
+
+  ProcessorState& state = encoding_state_;
+  Buffer::Instance& buffer = state.doDataList.push(data);
+  auto s = dynamic_cast<processState*>(&state);
+  state.setFilterState(FilterState::ProcessingData);
+
+  GoUint64 if_end_stream = dynamic_lib_->envoyGoEncodeData(
+    s, end_stream ? 1 : 0, reinterpret_cast<uint64_t>(&buffer), buffer.length());
+  if (if_end_stream == static_cast<GoUint64>(EndStreamType::NotEndStream)) {
+    end_stream = false;
+  } else {
+    end_stream = true;
+  }
+
+  state.setFilterState(FilterState::Done);
+  state.doDataList.moveOut(data);
+
+  upstream_conn_data_->connection().write(data, end_stream);
+}
+
+void TcpUpstream::doHeadersGo(ProcessorState& state, const Envoy::Http::RequestOrResponseHeaderMap& headers, bool end_stream) {
+  state.headers = &headers;
+  auto s = dynamic_cast<processState*>(&state);
+  state.setFilterState(FilterState::ProcessingData);
+
+  dynamic_lib_->envoyGoEncodeHeader(
+  s, end_stream ? 1 : 0, headers.size(),headers.byteSize());
+
+  }
+
+Envoy::Http::Status TcpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderMap& headers,
+                                               bool end_stream) {
+   
+  ENVOY_LOG(debug, "tcp upstream encodeHeaders: {}", headers);
+
+  initRequest();
+
+  ProcessorState& state = encoding_state_;
+  doHeadersGo(state, headers, end_stream);
```

Comment:
We'd better let golang return a data buffer, so that we can do this in Envoy C++ side.
```
upstream_conn_data_->connection().write(data, end_stream)
```

So that we can convert a HTTP GET request, without any body data, to any RPC request, thoughts?

Reply:
Good idea! already add buffer param in envoyGoEncodeHeader, to set buffer for writing upstream data when EncodeHeader in Go side.